In [2]:
#D1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.
import pandas as pd
import numpy as np
import os  

path= r'/Users/katep/08-2023 Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [5]:
#D1-2. Checked mixed-type data column.
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

   

product_name


In [19]:
#Cleaned mixed-type data column.
df_prods2=df_prods
df_prods2['product_name'] = df_prods2['product_name'].astype('str')


In [6]:
#D1-3. Checked missing values.
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [21]:
#Cleaned missing values.
df_prods2['product_name'].fillna('NA', inplace = True)

In [8]:
#D1-4. Checked duplicates.
df_dups = df_prods[df_prods.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
#Cleaned duplicates.
df_prods2_no_dup = df_prods2.drop_duplicates()

In [9]:
#D2. Run the df.describe() function on your df_ords dataframe. 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col=False)
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#D2-2. df_ords looks fine. Nothing's off. 

In [18]:
#D3. Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [12]:
#D5. Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#D5-2. The days_since_prior_order column has 206209 missing values. One of the reasons for the missing values could be due to the large number of first orders with no prior orders before them. 

In [26]:
#D6. Address the missing values using an appropriate method.
df_ords2=df_ords
df_ords2['missing_value_flag']=df_ords['days_since_prior_order'].isnull()

#D6-2. Why.
Create a new variable that acts like a flag based on the missing value.
Because there is a large number of missing values and there could be various reasons waiting upon further investigation. 

In [17]:
#7. Run a check for duplicate values in your df_ords data.
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#7-2. There is no duplicate value found in df_ords.

In [27]:
#Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.
df_prods2_no_dup.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))
df_ords2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))